### ISSUE TITLE
RoPE embeddings & Sinusoidal embeddings

### ISSUE THREAD

**Issue by Artur-Galstyan** [patrick-kidger/equinox#566]
Hey Patrick,

I was wondering if we could add RoPE and Sinusoidal embeddings to Equinox. It got a bit tedious to rewrite those every time, so I think it'd be nice to include it. 

I went ahead and provided 2 implementations: one with and one without the use of complex numbers.

Here's the complex numbers version:

```python
def precompute_freqs_cis(
    dim: int, end: int, theta: float = 10000.0
) -> Complex[Array, "end dim/2"]:
    def polar(abs, angle):
        return jnp.array(
            abs * jnp.cos(angle) + abs * jnp.sin(angle) * 1j, dtype=jnp.complex64
        )

    freqs = 1.0 / (theta ** (jnp.arange(0, dim, 2)[jnp.newaxis, :] / dim))
    t = jnp.arange(end)
    freqs_outer = jnp.outer(t, freqs)
    freqs_cis = polar(jnp.ones_like(freqs_outer), freqs_outer)
    return freqs_cis


def get_rope_embeddings_complex(x: Float[Array, "seq_len dim"]):
    def negate_half(x: Float[Array, "seq_len dim"]):
        d_2 = x.shape[-1] // 2
        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)

    seq_len, dim = x.shape
    freqs = precompute_freqs_cis(dim, seq_len)
    neg_half_x = negate_half(x)
    freqs_real = jnp.tile(freqs.real, (1, 2))
    freqs_imag = jnp.tile(freqs.imag, (1, 2))

    x_rope = (x * freqs_real) + (neg_half_x * freqs_imag)
    return x_rope
```

and this is the real numbers version (what you would fine in the RoFormer paper)
```python
def get_freqs(seq_len: int, dim: int, base: int = 10000):
    theta = 1 / (base ** (jnp.arange(0, dim, 2) / dim))
    t = jnp.arange(seq_len)

    idx_theta = jnp.einsum("i,j->ij", t, theta)
    idx_theta = jnp.concatenate([idx_theta, idx_theta], axis=1)

    freqs_cos = jnp.cos(idx_theta)
    freqs_sin = jnp.sin(idx_theta)

    return freqs_cos, freqs_sin


def get_rope_embedding(x: Float[Array, "seq_len dim"]):
    def negate_half(x: Float[Array, "seq_len dim"]):
        d_2 = x.shape[-1] // 2
        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)

    seq_len, dim = x.shape
    cos, sin = get_freqs(seq_len, dim)
    neg_half_x = negate_half(x)
    x_rope = (x * cos) + (neg_half_x * sin)
    return x_rope
```
Clearly, this is the not-cached version, i.e. you actually only need to compute the frequencies once and that keep those as a static field (I think as a static field).

I went ahead and performed a performance test on both:

`dim=1024` and `seq_len=4096` on MacBook Pro M1 Base model. Results are in seconds.

|          | Complex | Real      | 
|----------|---------|-----------|
| No JIT   | 299.1253| 193.1669  |
| Avg No Jit| 0.029912 | 0.01931 |
| With JIT | 71.4320 | 83.8433   |
| Avg With Jit| 0.0071 | 0.0083|

As you can see, the "real" version is *much* faster when not jitted, but the complex version is faster when you jit it.

My questions are:
1. Is there any interest in adding RoPE & Sinusoidal Embeddings?
2. If yes, which version should I add?

I think it'd be very helpful to have those as you frequently use those when implementing a transformer.

For reference, here's the entire code that you can copy paste and run (it also contains the sinusoidal embeddings too).

<details>
  <summary>Click me</summary>
  
  ```python
import sys
import time
from typing import Optional, Sequence, Union

import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np
from icecream import ic
from jaxtyping import Array, Complex, Float, Int, PRNGKeyArray
from matplotlib.animation import FuncAnimation
from matplotlib.patches import FancyArrowPatch

import equinox as eqx


# for reference, this is the ported version of the original code from LLaMA
# which I shamelessly stole, I mean borrowed from
def apply_rotary_emb(
    xq: Float[Array, "seq_len n_heads max_seq_len"],
    xk: Float[Array, "seq_len n_heads max_seq_len"],
    freqs_cis: Complex[Array, "seq_len dim//n_heads"],
) -> tuple[
    Float[Array, "seq_len n_heads max_seq_len"],
    Float[Array, "seq_len n_heads max_seq_len"],
]:
    def reshape_for_broadcast(freqs_cis: Array, x: Array):
        ndim = x.ndim
        assert 0 <= 1 < ndim
        assert freqs_cis.shape == (
            x.shape[1],
            x.shape[-1],
        ), f"{freqs_cis.shape=} != {(x.shape[1], x.shape[-1])}"
        shape = [d if i == 1 or i == ndim - 1 else 1 for i, d in enumerate(x.shape)]
        return freqs_cis.reshape(*shape)

    def reshape_complex_array(x: Array):
        return jax.lax.collapse(x, len(x.shape) - 2, len(x.shape))

    # e.g.: if xq.shape = 1, 32, 128, then xq_reshaped.shape =  1, 32, -1, 2 => 1, 32, 64, 2
    # it's 64 because total_elements = 1 * 32 * 128 = 4096 and 4096 / (1 * 32 * 2) = 64

    # also, because complex does not accept float16, we have to parse it to float32
    xq_reshaped = jnp.array(xq.reshape(*xq.shape[:-1], -1, 2), dtype=jnp.float32)
    xq_ = jax.lax.complex(
        xq_reshaped[..., 0], xq_reshaped[..., 1]
    )  # the last dimension is of size 2, the "..." skip over the first dimensions until we reach the last one
    xk_reshaped = jnp.array(xk.reshape(*xk.shape[:-1], -1, 2), dtype=jnp.float32)
    xk_ = jax.lax.complex(xk_reshaped[..., 0], xk_reshaped[..., 1])

    freqs_cis = reshape_for_broadcast(freqs_cis, xq_)
    tq = xq_ * freqs_cis
    tk = xk_ * freqs_cis
    xk_out = reshape_complex_array(jnp.stack((jnp.real(tk), jnp.imag(tk)), axis=-1))
    xq_out = reshape_complex_array(jnp.stack((jnp.real(tq), jnp.imag(tq)), axis=-1))
    return xq_out, xk_out


# standard positional encoding
def get_positional_encoding(
    n_tokens: int, n_vocab: int
) -> Float[Array, "n_tokens n_vocab"]:
    pos = jnp.arange(n_tokens)[:, jnp.newaxis]
    div_term = jnp.exp(jnp.arange(0, n_vocab, 2) * -(jnp.log(10000.0) / n_vocab))
    # the following expression is closer to the actual notation they used.
    # div_term = 1 / 10000 ** (jnp.arange(0, n_vocab, 2) / n_vocab)
    pos_enc = jnp.zeros((n_tokens, n_vocab))
    pos_enc = pos_enc.at[:, 0::2].set(jnp.sin(pos * div_term))
    pos_enc = pos_enc.at[:, 1::2].set(jnp.cos(pos * div_term))
    return pos_enc


def sinusoidal_animation():
    def update(num, x, pos_enc, arrow, label):
        updated = x + pos_enc[num]
        arrow.set_positions((0, 0), (updated[1], updated[0]))
        label.set_position((updated[1] + 0.1, updated[0] + 0.1))
        label.set_text(str(num))
        return arrow, label

    pos_enc = get_positional_encoding(8, 2)
    x = jnp.array([1, 1])

    xmin, xmax, ymin, ymax = -3, 3, -3, 3
    ticks_frequency = 1

    fig, ax = plt.subplots()
    ax.set_xlim(xmin, xmax)
    ax.set_ylim(ymin, ymax)
    ax.set(xlim=(xmin - 1, xmax + 1), ylim=(ymin - 1, ymax + 1), aspect="equal")

    ax.spines["bottom"].set_position("zero")
    ax.spines["left"].set_position("zero")

    # Remove top and right spines
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    # Create 'x' and 'y' labels placed at the end of the axes
    ax.set_xlabel("x", size=14, labelpad=-24, x=1.03)
    ax.set_ylabel("y", size=14, labelpad=-21, y=1.02, rotation=0)

    # Create custom major ticks to determine position of tick labels
    x_ticks = np.arange(xmin, xmax + 1, ticks_frequency)
    y_ticks = np.arange(ymin, ymax + 1, ticks_frequency)
    ax.set_xticks(x_ticks[x_ticks != 0])
    ax.set_yticks(y_ticks[y_ticks != 0])

    # Create minor ticks placed at each integer to enable drawing of minor grid
    # lines: note that this has no effect in this example with ticks_frequency=1
    ax.set_xticks(np.arange(xmin, xmax + 1), minor=True)
    ax.set_yticks(np.arange(ymin, ymax + 1), minor=True)

    # Draw major and minor grid lines
    ax.grid(which="both", color="grey", linewidth=1, linestyle="-", alpha=0.2)

    # Draw arrows
    arrow_fmt = dict(markersize=4, color="black", clip_on=False)
    ax.plot((1), (0), marker=">", transform=ax.get_yaxis_transform(), **arrow_fmt)
    ax.plot((0), (1), marker="^", transform=ax.get_xaxis_transform(), **arrow_fmt)

    # (line,) = ax.plot([], [], "r-")

    arrow = FancyArrowPatch((0, 0), (x[1], x[0]), mutation_scale=5, color="r")
    ax.add_patch(arrow)
    label = ax.text(x[1] + 0.1, x[0] + 0.1, "0")
    ani = FuncAnimation(
        fig,
        update,
        len(pos_enc),
        fargs=[x, pos_enc, arrow, label],
        interval=500,
        blit=True,
    )
    plt.show()


def get_freqs(seq_len: int, dim: int, base: int = 10000):
    theta = 1 / (base ** (jnp.arange(0, dim, 2) / dim))
    t = jnp.arange(seq_len)

    idx_theta = jnp.einsum("i,j->ij", t, theta)
    idx_theta = jnp.concatenate([idx_theta, idx_theta], axis=1)

    freqs_cos = jnp.cos(idx_theta)
    freqs_sin = jnp.sin(idx_theta)

    return freqs_cos, freqs_sin


def get_rope_embedding(x: Float[Array, "seq_len dim"]):
    def negate_half(x: Float[Array, "seq_len dim"]):
        d_2 = x.shape[-1] // 2
        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)

    seq_len, dim = x.shape
    cos, sin = get_freqs(seq_len, dim)
    neg_half_x = negate_half(x)
    x_rope = (x * cos) + (neg_half_x * sin)
    return x_rope


def precompute_freqs_cis(
    dim: int, end: int, theta: float = 10000.0
) -> Complex[Array, "end dim/2"]:
    def polar(abs, angle):
        return jnp.array(
            abs * jnp.cos(angle) + abs * jnp.sin(angle) * 1j, dtype=jnp.complex64
        )

    freqs = 1.0 / (theta ** (jnp.arange(0, dim, 2)[jnp.newaxis, :] / dim))
    t = jnp.arange(end)
    freqs_outer = jnp.outer(t, freqs)
    freqs_cis = polar(jnp.ones_like(freqs_outer), freqs_outer)
    return freqs_cis


def get_rope_embeddings_complex(x: Float[Array, "seq_len dim"]):
    def negate_half(x: Float[Array, "seq_len dim"]):
        d_2 = x.shape[-1] // 2
        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)

    seq_len, dim = x.shape
    freqs = precompute_freqs_cis(dim, seq_len)
    neg_half_x = negate_half(x)
    freqs_real = jnp.tile(freqs.real, (1, 2))
    freqs_imag = jnp.tile(freqs.imag, (1, 2))
    # freqs_real = freqs.real[..., jnp.newaxis]
    # freqs_imag = freqs.imag[..., jnp.newaxis]

    x_rope = (x * freqs_real) + (neg_half_x * freqs_imag)
    return x_rope


def rope_animation():
    def update(num, x, pos_enc, arrow, label):
        updated = pos_enc[num]
        arrow.set_positions((0, 0), (updated[1], updated[0]))
        label.set_position((updated[1] + 0.1, updated[0] + 0.1))
        label.set_text(str(num))
        return arrow, label

    x = jnp.ones((8, 2))
    pos_enc = get_rope_embedding(x)

    xmin, xmax, ymin, ymax = -3, 3, -3, 3
    ticks_frequency = 1

    fig, ax = plt.subplots()
    ax.set_xlim(xmin, xmax)
    ax.set_ylim(ymin, ymax)
    ax.set(xlim=(xmin - 1, xmax + 1), ylim=(ymin - 1, ymax + 1), aspect="equal")

    ax.spines["bottom"].set_position("zero")
    ax.spines["left"].set_position("zero")

    # Remove top and right spines
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    # Create 'x' and 'y' labels placed at the end of the axes
    ax.set_xlabel("x", size=14, labelpad=-24, x=1.03)
    ax.set_ylabel("y", size=14, labelpad=-21, y=1.02, rotation=0)

    # Create custom major ticks to determine position of tick labels
    x_ticks = np.arange(xmin, xmax + 1, ticks_frequency)
    y_ticks = np.arange(ymin, ymax + 1, ticks_frequency)
    ax.set_xticks(x_ticks[x_ticks != 0])
    ax.set_yticks(y_ticks[y_ticks != 0])

    # Create minor ticks placed at each integer to enable drawing of minor grid
    # lines: note that this has no effect in this example with ticks_frequency=1
    ax.set_xticks(np.arange(xmin, xmax + 1), minor=True)
    ax.set_yticks(np.arange(ymin, ymax + 1), minor=True)

    # Draw major and minor grid lines
    ax.grid(which="both", color="grey", linewidth=1, linestyle="-", alpha=0.2)

    # Draw arrows
    arrow_fmt = dict(markersize=4, color="black", clip_on=False)
    ax.plot((1), (0), marker=">", transform=ax.get_yaxis_transform(), **arrow_fmt)
    ax.plot((0), (1), marker="^", transform=ax.get_xaxis_transform(), **arrow_fmt)

    # (line,) = ax.plot([], [], "r-")

    arrow = FancyArrowPatch((0, 0), (0, 0), mutation_scale=5, color="r")
    ax.add_patch(arrow)
    label = ax.text(x[1] + 0.1, x[0] + 0.1, "0")
    ani = FuncAnimation(
        fig,
        update,
        len(pos_enc),
        fargs=[x, pos_enc, arrow, label],
        interval=1000,
        blit=True,
    )
    plt.show()


if __name__ == "__main__":
    dim = 1024
    max_seq_len = 4096

    x_rope_real = get_rope_embedding(jnp.ones((max_seq_len, dim)))
    x_rope_complex = get_rope_embeddings_complex(jnp.ones((max_seq_len, dim)))

    ic(x_rope_real)
    ic(x_rope_complex)

    ic(jnp.allclose(x_rope_real, x_rope_complex))
    # testing the rope embedding
    test_cases = 1_000_0

    x = jax.random.uniform(
        jax.random.PRNGKey(42), (max_seq_len, dim), dtype=jnp.float32
    )
    start_time = time.time()
    for _ in range(test_cases):
        x_rope_real = get_rope_embedding(x)
    end_time = time.time()
    ic("Real:", end_time - start_time)
    ic("Avg:", (end_time - start_time) / test_cases)
    start_time = time.time()
    for _ in range(test_cases):
        x_rope_complex = get_rope_embeddings_complex(x)
    end_time = time.time()
    ic("Complex: ", end_time - start_time)
    ic("Avg:", (end_time - start_time) / test_cases)

    # rope_animation()
    # sinusoidal_animation()

  ```
</details>

**Comment by patrick-kidger** [comment]
I definitely think it'd be worth getting these extra embeddings into Equinox. (And I've seen your PR on this as well.) I think for RoPE then it will mean also tweaking the `MultiheadAttention` implementation, to be able to insert the embeddings in the appropriate place. (Broadly speaking this is dovetailing with the job of improving our MHA implementation. E.g. I also intend to make it robust to low-precision training by casting to at least float32 for the softmax.)

Interesting to see the performance difference between real and complex, that's a detail I hadn't considered. I think we should focus on JIT performance, so probably complex makes most sense.

**Comment by Artur-Galstyan** [comment]
Indeed, I was thinking about adding an additional field in the MHA code, like `apply_rope_embeddings` and if set, it would apply RoPE to the query and key after the linear projections. 

Alternatively, it could also be a literal, like `apply_positional_embedding="rope" | "sinusoidal" | None` and if sinusoidal, it would apply it before the linear projections. 

What do you think?
